### ISSUE TITLE
Implement `__eq__` comparison for `TransformContainer` and `TransformProgram`

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#4824]
We have a new system of handling transforms that allows "programs" that apply a sequence of transforms during the execution pipeline. This allows for a more trackable, composable system.

This issue would be to implement `__eq__` comparison on `TransformContainer`:

https://github.com/PennyLaneAI/pennylane/blob/677fbe00f70e8456dacd5b0c4e9a46923036eeca/pennylane/transforms/core/transform_dispatcher.py#L330

and `TransformProgram`:

https://github.com/PennyLaneAI/pennylane/blob/677fbe00f70e8456dacd5b0c4e9a46923036eeca/pennylane/transforms/core/transform_program.py#L103

At the end of this, we should be able to do:

```
>>> from pennylane.transforms.core import TransformContainer, TransformProgram
>>> t1 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
>>> t2 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
>>> t3 = TransformContainer(qml.transforms.transpile.transform, kwargs={'coupling_map': [(0,1),(1,2)]})
>>> t1 == t2
True
>>> t2 == t3
False
>>> p1 = TransformProgram([t1,t3])
>>> p2 = TransformProgram([t2, t3])
>>> p3 == TransformProgram([t3, t2])
>>> p1 == p2
True
>>> p1 == p3
False
```

These comparisons can be implemented by overriding the `__eq__(self, other) -> bool` method of each class. Comparison with a different type of object should return `False`:

```
>>> t1 == 2
False
>>> p1 == t1
False
```

Tests can be added to:

https://github.com/PennyLaneAI/pennylane/blob/master/tests/transforms/test_experimental/test_transform_dispatcher.py

and 

https://github.com/PennyLaneAI/pennylane/blob/master/tests/transforms/test_experimental/test_transform_program.py

**Comment by AnuravModak** [comment]
Sure and thanks. I will start working on it.

**Comment by AnuravModak** [comment]
Now, while creating an object of type TransformContainer like this:

`t1 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})`
I am getting this error:
![image](https://github.com/PennyLaneAI/pennylane/assets/79191029/cf5ace98-a704-4891-af5e-208be9ce37ee)

So, if i am creating the same object according to the suggestion, like below, then its working fine:
`t1 = TransformContainer(qml.transforms.compile.transform_fn, kwargs={'num_passes': 2, 'expand_depth': 1})`

Are they identical? Kindly explain. If not, what should be done?

**Comment by albi3ro** [comment]
@AnuravModak Which version of PennyLane are you using?

In the last release, we switched `compile` from being a `qfunc_transform` to a new style using `transform` to create a `TransformDispatcher`.  You may be using a version of pennylane from before we made this change.

**Comment by AnuravModak** [comment]
I have one more doubt regarding p1 and p2 in your example above, what is their type?. Because when i try to implement something like this for class TransformProgram:
```
def __eq__(self, other):
        if not isinstance(other, TransformProgram):
            return False
        
        return(
            (self._transform_program == other._transform_program) &
            (self._classical_jacobians == other._classical_jacobians) &
            (self._argnums == other._argnums)
        )
```
Now, i have noticed that for the example below, i am getting False for both the cases from the isinstance() in the  `__eq__` method: whereas when i print the type of self or other this is returned `"<class '__main__.TransformProgram'>"` and type(TransformProgram) is `_program.py` which is a module. So what is that i am missing here? 

```
t1 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
t2 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
t3 = TransformContainer(qml.transforms.transpile.transform, kwargs={'coupling_map': [(0,1),(1,2)]})
p1 = TransformProgram([t1,t3])
p2 = TransformProgram([t2, t3])
p3 = TransformProgram([t3, t2])
print(p1==p2)
print(p1==p3)
```

please advise!

**Comment by albi3ro** [comment]
It's a little hard to comment without knowing the specifics, but the type should be:

```
>>> from pennylane.transforms.core import TransformProgram
>>> p = TransformProgram()
>>> type(p)
pennylane.transforms.core.transform_program.TransformProgram
```

Are you using a version of the class defined in a different location than the PennyLane source code? You may be having some namespacing issues.

You can see examples of how we override `__eq__` sometimes here: https://github.com/PennyLaneAI/pennylane/blob/3dee10187e0698d521daa841e5d146e92a50f343/pennylane/wires.py#L138

Feel free to open up a draft PR at any point

**Comment by AnuravModak** [comment]
Also, i want little confirmation here regarding p1 and p2
p1==p2 is True, is it because t1==t2 is True and t3==t3 is True.

If that's the case suppose i have this scenario below:

t1 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
t2 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
t3 = TransformContainer(qml.transforms.transpile.transform, kwargs={'coupling_map': [(0,1),(1,2)]})
t4 = TransformContainer(qml.transforms.compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
p1 = TransformProgram([t1,t3])
p2 = TransformProgram([t2, t3])
p3 = TransformProgram([t3, t2])
p4= TransformProgram([t1, t4])
p5= TransformProgram([t2,t1])
p4==p5
--> True 

is this right (True)? Or is it False because the equality comparison for TransformProgram depends on the order of the containers. 
Since the order of the containers is different between p4 and p5, the result of p4 == p5 is False.

And also confirm is this explanation right or not?

the equality comparison for TransformProgram takes into account both the equality of individual containers and their order. If the order of the containers is different, even if the containers themselves are equal, the programs are considered different. In my case, the order difference between t1, t4 and t2, t1 leads to p4 == p5 being False

Please advise.

**Comment by albi3ro** [comment]
So `t1` and `t2` have the same contents, so they should be equal. `t4` and `t1` also have the same contents and are thus equal.

So `[t1, t4]` and `[t2, t1]` should then be equal.  


But equality should depend on the order, just like how `[1,2]` is different from `[2,1]`.  

For example, we would have `[t1, t3]` be different than `[t3, t1]`.

Might help to give these transform better names?

```
from pennylane.transforms import compile, transpile

compile_1= TransformContainer(compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
compile_2 = TransformContainer(compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})
compiile_3= TransformContainer(compile.transform, kwargs={'num_passes': 2, 'expand_depth': 1})

transpile_1 = TransformContainer(transpile.transform, kwargs={'coupling_map': [(0,1),(1,2)]})
```

**Comment by albi3ro** [comment]
This issue has been closed by PR #4858
